// lyft-node/build/search/PriceEstimatesSearch.js
'use strict';
'use es6';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _immutable = require('immutable');

var _Coordinate = require('../data/Coordinate');

var _Coordinate2 = _interopRequireDefault(_Coordinate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var defaults = {
  start: new _Coordinate2.default(),
  end: new _Coordinate2.default(),
  seatCount: 2
};

var PriceEstimatesSearch = function (_Record) {
  _inherits(PriceEstimatesSearch, _Record);

  function PriceEstimatesSearch() {
    _classCallCheck(this, PriceEstimatesSearch);

    return _possibleConstructorReturn(this, (PriceEstimatesSearch.__proto__ || Object.getPrototypeOf(PriceEstimatesSearch)).apply(this, arguments));
  }

  _createClass(PriceEstimatesSearch, [{
    key: 'toParameters',
    value: function toParameters(search) {
      return (0, _immutable.Map)({
        start_latitude: this.start.latitude,
        start_longitude: this.start.longitude,
        end_latitude: this.end.latitude,
        end_longitude: this.end.longitude,
        seat_count: this.seatCount
      });
    }
  }]);

  return PriceEstimatesSearch;
}((0, _immutable.Record)(defaults));

exports.default = PriceEstimatesSearch;
